Editors/vscode/src/inlayHints.ts

```diff
@@ -0,0 +1,202 @@
+            }
+        }
+    })
+};
+
+interface SourceFile {
+    /** Source of the token for cancelling in-flight inlay hint requests. */
+    inFlightInlayHints: null | vscode.CancellationTokenSource;
+
+    /** Most recently applied decorations. */
+    cachedDecorations: null | vscode.DecorationOptions[];
+
+    /** The source file document in question. */
+    document: vscode.TextDocument;
+}
+
+class HintsUpdater implements vscode.Disposable {
+    private readonly disposables: vscode.Disposable[] = [];
+    private sourceFiles: Map<string, SourceFile> = new Map(); // uri -> SourceFile
+
+    constructor(private readonly client: langclient.LanguageClient) {
+        // Register listeners
+        vscode.window.onDidChangeVisibleTextEditors(this.onDidChangeVisibleTextEditors, this, this.disposables);
+        vscode.workspace.onDidChangeTextDocument(this.onDidChangeTextDocument, this, this.disposables);
+
+        // Set up initial cache
+        this.visibleSourceKitLSPEditors.forEach(editor => this.sourceFiles.set(
+            editor.document.uri.toString(),
+            {
+                document: editor.document,
+                inFlightInlayHints: null,
+                cachedDecorations: null
+            }
+        ));
+
+        this.syncCacheAndRenderHints();
+    }
+
+    private onDidChangeVisibleTextEditors(): void {
+        const newSourceFiles = new Map<string, SourceFile>();
+
+        // Rerender all, even up-to-date editors for simplicity
+        this.visibleSourceKitLSPEditors.forEach(async editor => {
+            const uri = editor.document.uri.toString();
+            const file = this.sourceFiles.get(uri) ?? {
+                document: editor.document,
+                inFlightInlayHints: null,
+                cachedDecorations: null
+            };
+            newSourceFiles.set(uri, file);
+
+            // No text documents changed, so we may try to use the cache
+            if (!file.cachedDecorations) {
+                const hints = await this.fetchHints(file);
+                if (!hints) return;
+                file.cachedDecorations = this.hintsToDecorations(hints);
+            }
+
+            this.renderDecorations(editor, file.cachedDecorations);
+        });
+
+        // Cancel requests for no longer visible (disposed) source files
+        this.sourceFiles.forEach((file, uri) => {
+            if (!newSourceFiles.has(uri)) {
+                file.inFlightInlayHints?.cancel();
+            }
+        });
+
+        this.sourceFiles = newSourceFiles;
+    }
+
+    private onDidChangeTextDocument(event: vscode.TextDocumentChangeEvent): void {
+        if (event.contentChanges.length !== 0 && this.isSourceKitLSPDocument(event.document)) {
+            this.syncCacheAndRenderHints();
+        }
+    }
+
+    private syncCacheAndRenderHints(): void {
+        this.sourceFiles.forEach(async (file, uri) => {
+            const hints = await this.fetchHints(file);
+            if (!hints) return;
```

Comment:
How would we end up in the case that `!hints`. IIUC `fetchHints` is returning `InlayHint[]` and JavaScript (thus probably also TypeScript) considers the empty array as boolean `true`.